cmd/catalog-importer/cmd/sync.go

```diff
@@ -240,33 +240,73 @@ func (opt *SyncOptions) Run(ctx context.Context, logger kitlog.Logger, cfg *conf
-			if opt.DryRun {
-				logger.Log("msg", "dry-run active, which means we fake a response")
-				updatedCatalogType = *catalogType // they start the same
-
-				// Then we pretend like we've already updated the schema, which means we rebuild the
-				// attributes.
-				updatedCatalogType.Schema = client.CatalogTypeSchemaV2{
-					Version:    updatedCatalogType.Schema.Version,
-					Attributes: []client.CatalogTypeAttributeV2{},
+	if opt.DryRun {
+		for _, outputType := range cfg.Outputs() {
+			baseModel, enumModels := output.MarshalType(outputType)
+			for _, model := range append(enumModels, baseModel) {
+				catalogType := catalogTypesByOutput[model.TypeName]
+
+				var updatedCatalogType client.CatalogTypeV2
+				if opt.DryRun {
+					logger.Log("msg", "dry-run active, which means we fake a response")
+					updatedCatalogType = *catalogType // they start the same
+
+					// Then we pretend like we've already updated the schema, which means we rebuild the
+					// attributes.
+					updatedCatalogType.Schema = client.CatalogTypeSchemaV2{
+						Version:    updatedCatalogType.Schema.Version,
+						Attributes: []client.CatalogTypeAttributeV2{},
+					}
+					for _, attr := range model.Attributes {
+						updatedCatalogType.Schema.Attributes = append(updatedCatalogType.Schema.Attributes, client.CatalogTypeAttributeV2{
+							Id:                *attr.Id,
+							Name:              attr.Name,
+							Type:              attr.Type,
+							Array:             attr.Array,
+							Mode:              client.CatalogTypeAttributeV2Mode(*attr.Mode),
+							BacklinkAttribute: attr.BacklinkAttribute,
+						})
+					}
 				}
+				OUT("  ✔ %s (id=%s)", model.TypeName, catalogType.Id)
+				DIFF("  ", *catalogType, updatedCatalogType)
+			}
+		}
+	} else {
+		// Update all the type schemas except for new backlinks, which could reference
+		// attributes that don't exist yet.
+		catalogTypeVersions := map[string]int64{}
+		for _, outputType := range cfg.Outputs() {
+			baseModel, enumModels := output.MarshalType(outputType)
+			for _, model := range append(enumModels, baseModel) {
+				catalogType := catalogTypesByOutput[model.TypeName]
+
+				attributesWithoutNewBacklinks := []client.CatalogTypeAttributePayloadV2{}
 				for _, attr := range model.Attributes {
-					updatedCatalogType.Schema.Attributes = append(updatedCatalogType.Schema.Attributes, client.CatalogTypeAttributeV2{
-						Id:    *attr.Id,
-						Name:  attr.Name,
-						Type:  attr.Type,
-						Array: attr.Array,
-					})
+					if attr.Mode != nil && *attr.Mode == client.CatalogTypeAttributePayloadV2ModeBacklink {
+						inCurrentSchema := false
+						// Does it exist in the current schema?
+						if attr.Id == nil {
+							logger.Log("msg", "attribute ID is empty, which shouldn't be possible")
+							continue
+						}
+
+						for _, existingAttr := range catalogType.Schema.Attributes {
+							if existingAttr.Id == *attr.Id {
+								inCurrentSchema = true
```

Comment:
If you just append directly here and whack the other append in an `else`, you can delete `inCurrentSchema` entirely and I think this whole thing becomes a lot more readable
```suggestion
								attributesWithoutNewBacklinks = append(attributesWithoutNewBacklinks, attr)
```

Reply:
ohh I see what you're saying. yes this is better, fixing.